luqum/utils.py

```diff
@@ -80,6 +80,128 @@ def __call__(self, tree):
         return self.visit(tree)
 
 
+class OpenRangeTransformer(visitor.TreeTransformer):
+    """Transforms open ranges to normal Range objects, i.e.
+
+    ::
+
+        >=foo            ->   [foo TO *]
+        <bar             ->   [* TO bar}
+
+    When *merge_ranges* is set (the default), this also merges open ranges in AND clauses::
+
+        >foo AND <=bar              ->   {foo TO bar]
+        [foo TO *] AND [* TO bar]   ->   [foo TO bar]
+
+    The merging of open ranges is performed by collecting all open ranges, and merging any
+    open range into any previously collected open range with an open bound. In other words, any
+    open bounds are always merged into the first suitable. Additionally, matching is always done
+    from left-to-right, so that this holds::
+
+        [a TO *] AND [b TO *] AND [* TO y] AND [* TO z]   ->   [a TO y] AND [b TO z]
+
+    Open ranges in OR and unknown clauses are not adjusted. Use :cls:`UnknownOperationResolver`
+    to make sure that unknown operations are resolved first.
+
+    Ranges with none of the bounds set are left unadjusted.
+    """
+
+    WILDCARD_WORD = Word("*")
+
+    def __init__(self, merge_ranges=True, add_head=" "):
+        self.merge_ranges = merge_ranges
+        self.add_head = add_head
+        super().__init__(track_parents=True)
+
+    def visit_and_operation(self, node, context):
+        if not self.merge_ranges:
+            yield from self.generic_visit(node, context)
+            return
+
+        new_node = AndOperation(pos=node.pos, size=node.size, head=node.head, tail=node.tail)
+        new_node.children = ()
+
+        # We collect all ranges from a AND's children with the same bound side, and pop the
+        # first one from the list when we encounter a different bound side. This allows us
+        # to join [x TO *] AND [y TO *] to [* TO a] AND [* TO b] as [x TO a] AND [y TO b]
+        # We adjust the first enounter of the range, and do not add the joined one.
+
+        # Note: possible_ranges_bound_side has no meaning if possible_ranges is empty
+        possible_ranges, possible_ranges_bound_side = [], None
```

Comment:
```suggestion
        possible_ranges = []
        # possible_ranges_bound_side is None if possible_ranges is empty.
        # After that, by construction, 
        # items in possible_ranges always have same bound_side, 
        # (because other nodes are merged until list is empty)
        possible_ranges_bound_side = None
```